docs: update creating source plugins guide & add example repo (#22943)
gatsbyjs/gatsby: examples/creating-source-plugins/README.md

# Creating First Class Gatsby Source Plugins

Create Gatsby plugins that leverage Gatsby's most impactful native features like remote image optimization, caching, customized GraphQL schemas and node relationships, and more.

> This monorepo serves as an example of a site using a first class source plugin to pull in data from a Node.js API. It is meant to show the 3 pieces that work together when building a source plugin, the API, the site, and the source plugin.

Reviewer: ```suggestion
This monorepo serves as an example of a site using a first class source plugin to pull in data from a Node.js API. It is meant to show the 3 pieces that work together when building a source plugin: the API, the site, and the source plugin.
```

## Setup

This monorepo uses yarn workspaces to manage the 3 indivdual projects:

- api: a Node.js API with in-memory data, and a Post and Author type, as well as support for subscriptions when Posts are mutated
- example-site: a barebones Gatsby site that implements the source plugin
- source-plugin: a plugin that uses several Gatsby APIs to source data from the API, create responsive/optimized images from remote locations, and link the nodes in the example site

To install dependencies for all projects run the install command in the root of the yarn workspace (which requires yarn to be installed):

```
yarn install
```

_Note: if you aren't using yarn, you can navigate into each of the 3 folders and run `npm install` instead_

Then you can run the api or example projects with:

> ```

Reviewer: The comments are confusing since this is shell. I'd remove them or break it out into two commands with text.

Author: That's a great point, I'll break them up.

// runs the api with yarn start at port 4000
yarn workspace api start
// runs the example site with gatsby develop at port 8000
yarn workspace example-site develop
```

> Running the example site also runs the plugin because it is included in the site's config. You'll see output in the console for different functionality, and then can open up the browser to localhost:8000 to see the site.

Reviewer: ```suggestion
Running the example site also runs the plugin because it is included in the site's config. You'll see output in the console for different functionality and then can open up the browser to `localhost:8000` to see the site.
```

## Developing and Experimenting

You can open up `localhost:4000` when the api is running and test a query like this to see data returned:

```graphql
{
  posts {
    id
  }
}
```

> You can also run 3 different mutations: `createPost`, `updatePost`, and `deletePost`. These methods would mimic CRUD operations happening on the API of the data source like a headless CMS.

Reviewer: How do you run the mutation? Code change? Flag? other?

Author: This happens inside of the GraphQL Playground opened up at `localhost:4000`, I'll add that in as context! 🙂 

```suggestion
You can also run 3 different mutations from the GraphQL Playground (at `localhost:4000`): `createPost`, `updatePost`, and `deletePost`. These methods would mimic CRUD operations happening on the API of the data source like a headless CMS. An example `updatePost` mutation is outlined below.
```

When you run a mutation on a post, a subscription event is published, which lets the plugin know it should respond and update nodes:

```graphql
mutation {
  updatePost(id: "post-id", description: "Some data!") {
    id
    slug
    description
  }
}
```

> The websites homepage will update to any changes while the source plugin is subscribed to changes, which is when the `preview: true` is provided in the example site's `gatsby-config`.

Reviewer: ```suggestion
The website's homepage will update with any changes while the source plugin is subscribed to changes, which is when the `preview: true` is provided in the example site's `gatsby-config`.
```

- source-plugin: a plugin that uses several Gatsby APIs to source data from the API, create responsive/optimized images from remote locations, and link the nodes in the example site

To install dependencies for all projects run the install command in the root of the yarn workspace (which requires yarn to be installed):

```
yarn install
```

_Note: if you aren't using yarn, you can navigate into each of the 3 folders and run `npm install` instead_

Then you can run the api or example projects in separate terminal windows with the commands below.

For the API which runs at `localhost:4000`, use this command:

```
yarn workspace api start
```

And to run the example site with `gatsby develop` at `localhost:8000`, use this command:

```
yarn workspace example-site develop
```

Running the example site also runs the plugin because it is included in the site's config. You'll see output in the console for different functionality and then can open up the browser to `localhost:8000` to see the site.

## Developing and Experimenting

You can open up `localhost:4000` when the api is running and test a query like this to see data returned:

```graphql
{
  posts {
    id
  }
}
```

> You can also run 3 different mutations from the GraphQL Playground (at `localhost:4000`): `createPost`, `updatePost`, and `deletePost`. These methods would mimic CRUD operations happening on the API of the data source like a headless CMS. An example `updatePost` mutation is outlined below.

Reviewer: Maybe worth adding a note saying that it's different than the typical graphiql endpoint of ___graphql?

Author: Good point, since the audience may have only been exposed to the ___graphiql endpoint, great feedback!

When you run a mutation on a post, a subscription event is published, which lets the plugin know it should respond and update nodes:

> ```graphql

Reviewer: Worth calling out that you need to update the id and data in this example, rather than copy/paste.
And how to get the id in the most straightforward way? I was able to get it with graphiql and a query, but it doesn't seem to be correct since I get errors when trying to run this example.

Author: Updated this and suggested copying a value from the query above it.

mutation {
  updatePost(id: "post-id", description: "Some data!") {
    id
    slug
    description
  }
}
```

The website's homepage will update with any changes while the source plugin is subscribed to changes, which is when the `preview: true` is provided in the example site's `gatsby-config`.

> You can also optionally listen for subscription events with this query in the playground which will display data when a mutatioin is run:

Reviewer: ```suggestion
You can also optionally listen for subscription events with this query in the playground which will display data when a mutation is run:
```
